### ISSUE TITLE
Reverse DNS app IDs are often a burden

### ISSUE THREAD

**Issue by allanday** [flatpak/flatpak#1258]
There's been some discussion about this on IRC today, and I thought that it would be good to have a more permanent place for that conversation.

To put it bluntly: I think that reverse DNS app IDs are inhibiting flatpak adoption. They are inherently more difficult to use than the shorter IDs that people are used to with package managers, as well as those used by Snaps. In comparison, Flatpak feels like a step backwards.

Reverse DNS paths are also unfamiliar to many users, meaning that they make the learning curve steeper, and can sometimes lead to confusion.

While I'm sure that there are good technical reasons for reverse DNS IDs, I can't help but wonder what purpose they serve in the majority of cases, where duplicate apps aren't present.

As a side note, I do think that it's important to have a good CLI experience for flatpak - it's important for developers and a lot of Linux users too, and these are people that we need to win over.

**Comment by matthiasclasen** [comment]
There was discussion on irc, but you only recorded your doubts here, not the answers you got :-)

Reverse dns ids are guaranteeing uniqueness and avoid conflicts between apps, which is particularly important with multiple independent repositories.

Reverse dns ids are also not a flatpak invention - they are d-bus names, and when we decided to unify desktop file name == bus-name == app id, we had to respect the naming rules for bus names.

They were never intended to be a primary end-user-visible identification for applications.

You are trying to judge the commandline interface by rules that it was never meant to satisfy.

**Comment by allanday** [comment]
> There was discussion on irc, but you only recorded your doubts here, not the answers you got :-)

I thought it was best for people to provide their own thoughts, rather than me trying to interpret them for them.

> Reverse dns ids are guaranteeing uniqueness and avoid conflicts between apps, which is particularly important with multiple independent repositories.

Other systems allow multiple independent repositories, and they don't have reverse DNS IDs. Also, I'm unsure whether optimising for multiple repositories containing duplicate apps is the best strategy - the best UX is when people have a single repository that provides all their apps.

> Reverse dns ids are also not a flatpak invention - they are d-bus names, and when we decided to unify desktop file name == bus-name == app id, we had to respect the naming rules for bus names.

I don't think we can or should assume familiarity with D-Bus.

> They were never intended to be a primary end-user-visible identification for applications.
>
> You are trying to judge the commandline interface by rules that it was never meant to satisfy.

I absolutely agree with the idea of using flatpak to power GUI tools. However, that doesn't mean that the CLI is irrelevant. A lot of people use the CLI, and a lot of those people are users and developers who we need to win over. Right now we are offering a CLI that is harder to use than that of our competitors, and I think that's harming us.

**Comment by julianrichen** [comment]
A compromise would be to have flatpak assume that if the argument after `flatpak install` has no `.` in it that it should just search for the last part of the reverse DNS name. So for example doing `flatpak install gedit` would resolve to `flatpak install org.gnome.gedit`. The chances that another application is called `gedit` is low.

However if users try to install say `flatpak install music` the output could be similar to when trying to install an application found in two repos:

```
$ flatpak install music

Found 2 applications matching `music`:

    [1] org.gnome.music
    [2] org.kde.music

Which would you like to install [1 - 2]: 
```

However, this assumes the average user would be using the CLI install of a front-end app, i.e. gnome-software. I image most CLI users wouldn't be scared away from reverse DNS app names.

**Comment by TingPing** [comment]
> Other systems allow multiple independent repositories, and they don't have reverse DNS IDs.

And conflicts are easier to happen there and security is worse (no way to enforce dbus/desktop/icon/etc ownership).

> I don't think we can or should assume familiarity with D-Bus.

It isn't about familiarity its a core part of the design, exported data files **must** match the app-id.

> A lot of people use the CLI, and a lot of those people are users and developers who we need to win over. Right now we are offering a CLI that is harder to use than that of our competitors, and I think that's harming us.

I do agree with that, thus why I implemented `search`, but the app-id is just one concept that cannot be avoided.


> A compromise would be to have flatpak assume that if the argument after flatpak install has no . in it that it should just search for the last part of the reverse DNS name. So for example doing flatpak install gedit would resolve to flatpak install org.gnome.gedit. The chances that another application is called gedit is low.

Probably search the full app-id (`com.skype.Client` for example) plus search the appdata name.

**Comment by allanday** [comment]
> I image most CLI users wouldn't be scared away from reverse DNS app names.

It's not a matter of whether someone is scared or not. People trying flatpak, including app developers, are often evaluating it alongside other options. In that situation it is rational to consider how much work each tool requires to learn and use.

**Comment by allanday** [comment]
> Other systems allow multiple independent repositories, and they don't have reverse DNS IDs.
> 
> And conflicts are easier to happen there and security is worse (no way to enforce dbus/desktop/icon/etc ownership).
> 
>     I don't think we can or should assume familiarity with D-Bus.
> 
> It isn't about familiarity its a core part of the design, exported data files must match the app-id.

If this is the case, one thing we could do is try and sell reverse DNS IDs as a feature - explain that they demonstrate how well-design flatpak is, and how they make it more secure. It's not an answer as such, but it could help.

**Comment by Grumbel** [comment]
For me as a developer the biggest problem with reverse DNS is simply that I don't own any DNS names for my smaller projects to begin with. So what am I going to use as app-id? Hoster URL, email, made up fantasy name? None of them is pretty. Hosting changes every few years, so do email addresses. Lot's of that happens naturally when a project growth.

From my perspective reverse-DNS names solve a problem (name collision) that is *extremely rare* in practice, while creating a ton of problems that are extremely common. It doesn't even make sense to use DNS-names when the system doesn't even do anything with the actual DNS system, it's basically just a cargo cult. And yes, Dbus does it too, but it sure as hell hasn't helped there, it just made method names and paths obnoxiously long.

Finally in the context of the current user interface there isn't even a need for a unique name, as it's `flatpak install REMOTE REF`, so the REMOTE already provides a namespace to avoid collisions.

If it absolutely has to be unique, use a UUID. If the user has to interact with it, use the actual name of the application and provide disambiguation functionality. Reverse-DNS is just an ugly middle ground.

**Comment by TingPing** [comment]
> Finally in the context of the current user interface there isn't even a need for a unique name, as it's flatpak install REMOTE REF, so the REMOTE already provides a namespace to avoid collisions.

That doesn't matter, all apps are installed into the system and that is where they conflict.

**Comment by allanday** [comment]
I've recently been experiencing this issue with regards to launching apps from the commandline. This is something I'll sometimes do if I'm manipulating a file using a CLI tool, and then want to open it with a graphical app. Doing this with flatpak feels hard work and awkward.

**Comment by aiamuzz** [comment]
ok ... i agree ... hence opened #2005 ...

[penned a few thoughts to consider](https://github.com/flatpak/flatpak/issues/2005#issuecomment-415575270) ...

**Comment by TingPing** [comment]
@aiamuzz `flatpak search` is helpful.

**Comment by aiamuzz** [comment]
@TingPing  ... hmmm ... a good stop-gap option ... until this is sorted out ... in an elegant way of course

**Comment by mwleeds** [comment]
After #2113 and #2288 were merged, we essentially have the behavior suggested in https://github.com/flatpak/flatpak/issues/1258#issuecomment-352192179. Here's what it looks like:
```
$ flatpak install music
Multiple remotes found with refs similar to ‘music’:
1) flathub (system)
Which do you want to use (0 to abort)? [0-1]: 1
Similar refs found for ‘music’ in remote ‘flathub’ (system):
1) app/com.googleplaymusicdesktopplayer.GPMDP/x86_64/stable
2) app/org.musicbrainz.Picard/x86_64/stable
3) app/com.github.artemanufrij.playmymusic/x86_64/stable
4) app/org.gnome.Music/x86_64/stable
Which do you want to use (0 to abort)? [0-4]: 4
Required runtime for org.gnome.Music/x86_64/stable (runtime/org.gnome.Platform/x86_64/3.30) found in remote flathub
Do you want to install it? [Y/n]: 
Installing in system:
org.gnome.Platform/x86_64/3.30        flathub f6b4bc140bb2
org.gnome.Platform.Locale/x86_64/3.30 flathub cd6fc7365d00
org.gnome.Music/x86_64/stable         flathub 1c88a72a61b1
  permissions: ipc, network, pulseaudio, wayland, x11
  file access: xdg-cache/media-art, xdg-music, xdg-run/dconf, ~/.config/dconf:ro
  dbus access: ca.desrt.dconf, com.intel.dleyna-server, org.freedesktop.Tracker1, org.gnome.OnlineAccounts, org.gnome.SettingsDaemon.MediaKeys, org.gtk.vfs.mountpoint_http
  dbus ownership: org.mpris.MediaPlayer2.GnomeMusic
org.gnome.Music.Locale/x86_64/stable  flathub 03c1e3dd6340
Updating in system:
org.freedesktop.Platform.html5-codecs/x86_64/18.08 flathub 0e0cea6e038f
Is this ok [Y/n]: 
Installing: org.gnome.Platform/x86_64/3.30 from flathub
[####################] 280 metadata, 3140 content objects fetched; 188853 KiB transferred in 20 seconds
Now at f6b4bc140bb2.
Installing: org.gnome.Platform.Locale/x86_64/3.30 from flathub
[####################] 3 metadata, 1 content objects fetched; 16 KiB transferred in 0 seconds
Now at cd6fc7365d00.
Updating: org.freedesktop.Platform.html5-codecs/x86_64/18.08 from flathub
[####################] 1 delta parts, 1 loose fetched; 637 B transferred in 0 seconds
Now at 0e0cea6e038f.
Installing: org.gnome.Music/x86_64/stable from flathub
[####################] 123 metadata, 328 content objects fetched; 2330 KiB transferred in 16 seconds
Now at 1c88a72a61b1.
Installing: org.gnome.Music.Locale/x86_64/stable from flathub
[####################] 3 metadata, 1 content objects fetched; 6 KiB transferred in 0 seconds
Now at 03c1e3dd6340.
```

You can also use `-y` to avoid some prompts if they only have one option:
```
$ flatpak install -y dbeaver
Required runtime for io.dbeaver.DBeaverCommunity/x86_64/stable (runtime/org.freedesktop.Platform/x86_64/18.08) found in remote flathub
Installing in system:
org.freedesktop.Platform/x86_64/18.08        flathub 95881b6e72a7
org.freedesktop.Platform.Locale/x86_64/18.08 flathub be0744bd1e50
io.dbeaver.DBeaverCommunity/x86_64/stable    flathub 04bee3ab61b3
  permissions: ipc, network, pulseaudio, wayland, x11, dri
  file access: home
  dbus access: org.freedesktop.Notifications
Installing: org.freedesktop.Platform/x86_64/18.08 from flathub
[####################] 119 metadata, 474 content objects fetched; 19980 KiB transferred in 3 seconds
Now at 95881b6e72a7.
Installing: org.freedesktop.Platform.Locale/x86_64/18.08 from flathub
[####################] 3 metadata, 1 content objects fetched; 16 KiB transferred in 0 seconds
Now at be0744bd1e50.
Installing: io.dbeaver.DBeaverCommunity/x86_64/stable from flathub
[####################] 111 metadata, 596 content objects fetched; 114733 KiB transferred in 11 seconds
Now at 04bee3ab61b3.
```

So should we close this issue or do we want to leave it open and add fuzzy-matching to other Flatpak commands?

**Comment by aiamuzz** [comment]
> So should we close this issue or do we want to leave it open and add fuzzy-matching to other Flatpak commands?

I guess leaving it open would be good idea ... as we can test and report back here ... as soon as we receive the update ...

**Comment by mwleeds** [comment]
No, only the install command as things stand now. It would probably make sense to add it to the uninstall command for symmetry.

**Comment by aiamuzz** [comment]
> No, only the install command as things stand now. It would probably make sense to add it to the uninstall command for symmetry.

@mwleeds  ... ok ...
I think even adding it to the 'run' command makes sense ... doesn't it ? The whole point of this implementation is to not have to remember the reverse DNS names ... so 'run' too requires this implementation.

i don't know if there are any other commands that requires one to type the whole package name ... i can only think of only 'Install', 'Uninstall' & 'Run' ... as a normal user ... any other use case scenario i may not be aware of ...

BTW waiting to test this new implementation ... as soon as it is released ...

**Comment by matthiasclasen** [comment]
I'd rather close this issue, tbh.

If you want to see this in other commands, send pull requests. It is just a matter of copying what Matthew did for install

**Comment by mwleeds** [comment]
Yes, I agree it makes sense to add fuzzy matching to more commands. I just haven't done it yet

**Comment by aiamuzz** [comment]
@matthiasclasen ... @mwleeds ... this matching option isn't available with `flatpak run` command ... it makes as much sense having it here as much as having it with the install command ... When the whole idea is not having to remember the `reverse.dns.app.names` ...

**Comment by mwleeds** [comment]
I agree we should add fuzzy ID matching to more commands. We can re-open this until that's done.

**Comment by matkoniecz** [comment]
> They were never intended to be a primary end-user-visible identification for applications.

Then how I am supposed to start flatpak installed software from the command line?

**Comment by matthiasclasen** [comment]
@mwleeds since you are currently revisiting cli wrappers, I recently wondered: We never added file forwarding to them, was there a reason for that, or ... we just never got around to it ?

**Comment by TingPing** [comment]
> We never added file forwarding to them, was there a reason for that, or ... we just never got around to it ?

I mentioned this elsewhere but just to document it.

The way `--file-forwarding` works is it looks for a specific identifier to know what is and isn't a forwardable URL or path. This works well for `.desktop` files because they use `@f` or `@u` for this.

Arbitrary CLI commands won't include this. So the options are limited; You can try to forward everything that looks file-ish which could be error prone or unwanted or you could require the user explicitly add the identifiers which is barely an improvement in UX.

**Comment by pothos** [comment]
Debian has [`update-alternatives`](https://man7.org/linux/man-pages/man1/update-alternatives.1.html) to manage symlinks in the `PATH` when multiple packages provide the same command.
Flatpak could also have such a system to let apps register a command and by the first app to be installed will be the default backing app for this command. The user would have to do the switch manually.
The symlinks could be in `~/.local/share/flatpak/exports/bin`, too and I think it would be good to have this in the `PATH` for every user by providing a `/etc/profile.d/flatpak-export.sh` script that adds it to the `PATH`.
To avoid conflicts with the system binaries the symlinks could get a prefix like `fp-` but one could also rely on the ordering of `PATH` to ensure that the symlinks are lowest priority (explicitly running a Flatpak would be with `flatpak run`).

**Comment by smcv** [comment]
This is already at the point where it's basically a "frequently asked question".

> Debian has [update-alternatives](https://man7.org/linux/man-pages/man1/update-alternatives.1.html) to manage symlinks in the PATH when multiple packages provide the same command.
Flatpak could also have such a system to let apps register a command and by the first app to be installed will be the default backing app for this command.

That doesn't prevent a malicious app from providing a command that the user expects to trust with secret or otherwise important information, like for example `pass` (a password manager) or `gpg`, or [typosquatting](https://en.wikipedia.org/wiki/Typosquatting) a sufficiently-similar command like `psss`.

>  To avoid conflicts with the system binaries the symlinks could get a prefix like fp-

That doesn't prevent a malicious app from "stealing" a command name from a different, more-trusted Flatpak app, either directly or by typosquatting a very similar name. For example, a malicious or compromised game could export `fp-seahorse` in the hope that it gets installed before [Seahorse](https://flathub.org/apps/details/org.gnome.seahorse.Application) and can "steal" its CLI.

> but one could also rely on the ordering of PATH to ensure that the symlinks are lowest priority (explicitly running a Flatpak would be with flatpak run)

Again, that doesn't prevent a malicious app from "stealing" a command name from a different, more-trusted Flatpak app, or typosquatting a name that is very similar to either a system executable or another Flatpak app.

The Unix, interactive-shell-prompt, command pipeline design tradition is very powerful, and can be very concise when you fully and equally trust all the applications (and you never make mistakes), but the other side of it being that powerful is that it can go horribly wrong if there are components that are actively working to undermine you. Flatpak's security model is that apps should be meaningfully sandboxed, and if some apps are malicious or compromised, that shouldn't be a full system compromise.

(Obviously some Flatpak permissions make the app that has those permissions very powerful, and if an app needs those permissions, then that specific app being malicious or compromised would be game over: your only choices are to trust it or not run it. The goal is to reduce the number of those apps over time, ideally down to zero.)

Luckily, the Unix shell environment is fully programmable. If you configure your shell with something like `alias inkscape="flatpak run org.inkscape.Inkscape"` or `inkscape () { flatpak run org.inkscape.Inkscape "$@" }` for the minority of Flatpak apps that you expect to run from a CLI, that sidesteps all of these concerns by incorporating the app name that *you* intended to be identified as `inkscape`.

**Comment by matkoniecz** [comment]
> Luckily, the Unix shell environment is fully programmable. If you configure your shell with something like alias inkscape="flatpak run org.inkscape.Inkscape" or inkscape () { flatpak run org.inkscape.Inkscape "$@" } for the minority of Flatpak apps that you expect to run from a CLI, that sidesteps all of these concerns by incorporating the app name that you intended to be identified as inkscape.

Making it reasonably easy to automate it without manual fiddling with dotfiles (extra option on install?) would make flatpak far more usable.

or at least provide extra info in install instructions

**Comment by Ferk** [comment]
> That doesn't prevent a malicious app from "stealing" a command name from a different, more-trusted Flatpak app, either directly or by typosquatting a very similar name.

Does flatpak currently offer a real defense against this? (eg. typosquatting `com.gnome`, `org.gnomme` or something like that).

Also, isn't it already the case that a malicious flatpak could have the same id as another, from a different malicious source/repository, and get installed before the original flatpak?

I don't know if  things have changed, but last time I checked the `org.gimp.GIMP` you can install from fedora sources is a different GIMP from the one you can install from flathub. It was even built against different runtimes.

> Luckily, the Unix shell environment is fully programmable. If you configure your shell with something like alias inkscape="flatpak run org.inkscape.Inkscape" or inkscape () { flatpak run org.inkscape.Inkscape "$@" } for the minority of Flatpak apps that you expect to run from a CLI, that sidesteps all of these concerns by incorporating the app name that you intended to be identified as inkscape.

If you assume that the Flatpak apps you'll run from the `$PATH` is a minority and that the GUI launchers everyone uses will all have support for flatpak or `.desktop` files always, then you might not need to automate it. But then the CLI usability is just one more thing that a proper installation of inkscape has that flatpak doesn't. This is just one reason why flatpak is better suited for those who don't really use the CLI all that much for running their apps, but might be more annoying to use for those who do.

**Comment by dead10ck** [comment]
>That doesn't prevent a malicious app from providing a command that the user expects to trust with secret or otherwise important information, like for example pass (a password manager) or gpg, or [typosquatting](https://en.wikipedia.org/wiki/Typosquatting) a sufficiently-similar command like psss.

At this rate, we may as well stop using word processing software because someone might send us a malicious word doc with an innocuous name. Or stop using video player software because someone could send you a file named `happy-puppies.mp4` that actually turns out to be porn. Might as well stop using DNS too because sometimes people typo squat domains like `pyapal.com`. Maybe we should disallow hyperlinks too because people can embed malicious links in them. Etc, etc, etc.

**Comment by pothos** [comment]
Flatpak apps have already ways of opting out of the sandbox. The permission system or even an interactive portal could also be used to let the app provide a CLI command.

**Comment by rokejulianlockhart** [comment]
I agree, @Obsessee, that `winget`'s style is much more sensical.

Periods are used to denote hierarchy within software development paradigms. That's obviously why DNS chose to use it, but I see no reason to duplicate specifically DNS's TLDs when dot-hierarchy is used in contexts without it.

It's not as if pretending that TLDs exist provides superior consistency:

https://learn.microsoft.com/en-us/dotnet/core/tutorials/with-visual-studio-code?pivots=dotnet-7-0

```c#
Console.WriteLine("What is your name?");
var name = Console.ReadLine();
var currentDate = DateTime.Now;
Console.WriteLine($"{Environment.NewLine}Hello, {name}, on {currentDate:d} at {currentDate:t}!");
Console.Write($"{Environment.NewLine}Press any key to exit...");
Console.ReadKey(true);
```

https://www.programiz.com/python-programming/modules

```python3
import math
print(math.pi)
```

Consequently, why should `flatpak` be the sole exception, especially with the exception being a strange obsession with TLD replication in a non-network context?

**Comment by Arcitec** [comment]
> > We never added file forwarding to them, was there a reason for that, or ... we just never got around to it ?
> 
> I mentioned this elsewhere but just to document it.
> 
> The way `--file-forwarding` works is it looks for a specific identifier to know what is and isn't a forwardable URL or path. This works well for `.desktop` files because they use `@f` or `@u` for this.
> 
> Arbitrary CLI commands don't have this information so the options are limited; You can try to forward everything that looks file-ish which could be error-prone/unwanted or you could require the user explicitly add the identifiers which is barely an improvement in UX.

The way I handle this in my own wrapper scripts is definitely less secure but works out for me:

- I only support apps that have `--filesystem=host`.
- There's a wrapper script, and a 2nd script that creates new wrappers (aliases).
- Usage is `fp-shim-create meld org.gnome.Meld`, and then simply `meld <whatever paths>`. The shim does the difficult job of translating paths to what the Flatpak will see, so that the sandboxed app can find the arguments that were given on the command line.
- **Thoughts for the official CLI alias feature for the future:** Having a way to map all "file-like arguments" to actual files somehow would be optimal, so that the `--filesystem=host` and manual mapping won't be needed anymore. We could simply have a Flatpak capability: "Map CLI filenames", which can be modified via overrides (Flatseal), and is always (?) off by default. It would then translate all file-like arguments into actual paths when using Flatpak's own official alias/shim.

Without further ado, this is how I "solved" the issue locally for now...

## fp-shim-wrapper

```sh
#!/usr/bin/env bash

# Mappings between host's / (root) paths and the sandbox. Assumes that Flatpak
# has "filesystem=host" permissions, which is obviously required for CLI usage.
# IMPORTANT: We will NOT translate complicated "/../" traversal. It's too complex,
# since we cannot be sure that the Flatpak's arguments should be treated as paths.
declare -A _SANDBOX_PATH_MAP=(
    ["/bin"]="/run/host/bin"
    ["/etc"]="/run/host/etc"
    ["/lib"]="/run/host/lib"
    ["/lib64"]="/run/host/lib64"
    ["/sbin"]="/run/host/sbin"
    ["/usr"]="/run/host/usr"
)

convert_path() {
    _CONVERTED_PATH="${1:-}"
    # If it begins with a slash, treat as possible path and apply conversion.
    if [[ "${_CONVERTED_PATH}" =~ ^(/[^/]+)(.*)$ ]]; then
        # Check if the path began with a directory that we should remap.
        if [[ -v _SANDBOX_PATH_MAP["${BASH_REMATCH[1]}"] ]]; then
            _CONVERTED_PATH="${_SANDBOX_PATH_MAP["${BASH_REMATCH[1]}"]}${BASH_REMATCH[2]}"
        fi
    fi
}

# First argument: The name of the flatpak.
flatpak_name="${1:-}"
shift 1

# Everything else: Arguments to pass to the Flatpak (allowed to be empty).
declare -a args=()
for arg in "$@"; do
    convert_path "${arg}"
    args+=("${_CONVERTED_PATH}")
done

# Translate the current working directory too (otherwise the shim fails when
# attempting to open relative file paths while inside special dirs on the host).
convert_path "${PWD}"
flatpak_cwd="${_CONVERTED_PATH}"

# Now execute the Flatpak with the custom working dir and remapped arguments.
flatpak run --cwd="${flatpak_cwd}" "${flatpak_name}" -- "${args[@]}"
```

## fp-shim-create

```sh
#!/usr/bin/env bash

set -oue pipefail

yell() { echo "${0}: ${*}"; }

shim_bin="${1:-}"
flatpak_id="${2:-}"

if [[ -z "${shim_bin}" ]] || [[ -z "${flatpak_id}" ]]; then
    yell "Usage Example: ${0} /usr/bin/example com.example.Example"
    exit 1
fi

yell "Creating \"${shim_bin}\" shim for \"${flatpak_id}\"."
cat > "${shim_bin}" << EOF
#!/usr/bin/env bash

"/usr/bin/fp-shim-wrapper" ${flatpak_id} "\${@}"
EOF

chmod +x "${shim_bin}"
```

**Comment by tutacat** [comment]
Just add your own symbolic links to /usr/share/flatpak/exports/bin/

**Comment by leandrokemp** [comment]
Running flatpaks without reversed DNS can be achieved with the suggested changes below:

1. When running "flatpak install", get the "command" value from the app's flatpak manifest
2. Create symlinks with the value retrieved from #1 in /var/lib/flatpak/exports/bin pointing to its corresponding reverse DNS binary
3. Append /var/lib/flatpak/exports/bin to $PATH

Note: If there is another flatpak app installed with the same "command", then prompt a message asking the user which reverse DNS (manifest app-id value) should be used as "command" in the terminal.

![image](https://github.com/user-attachments/assets/1ea040cf-d79b-4c4c-8447-91aa5bb4669e)

**Comment by razzeee** [comment]
Some examples of unwanted behavior with that proposed change

https://github.com/flathub/net.stabyourself.nottetris2/blob/5d9ca336d81947ab2e2d2989e35d5707d8c56dfa/net.stabyourself.nottetris2.yml#L5

https://github.com/flathub/io.gdevs.GDLauncher/blob/6a84fb1dc9ca409161b8c1fdc50c24c4fdf27915/io.gdevs.GDLauncher.yml#L7

https://github.com/flathub/org.srb2.SRB2Kart-Saturn/blob/bfa959fdc4596364da0e1607f434fb0f63aa0c14/org.srb2.SRB2Kart-Saturn.yaml#L5

https://github.com/flathub/io.github.freedoom.Phase2/blob/65f6dd3e3263d444db27caa30aefcb7595c96aa1/io.github.freedoom.Phase2.yaml#L5

https://github.com/flathub/org.sil.dictionary-app-builder/blob/df5ed7290a67fe5f499717c7156703076fbd715c/org.sil.dictionary-app-builder.yml#L27

**Comment by RokeJulianLockhart** [comment]
> https://github.com/flatpak/flatpak/issues/1258#issuecomment-2569067466

@razzeee, what's undesirable about those? Each file appears to have a name that correlates (at least, tangentially) to its accompanying rDNS ID. An example is [`/app/bin/nottetris2.sh`](https://github.com/flathub/net.stabyourself.nottetris2/blob/5d9ca336d81947ab2e2d2989e35d5707d8c56dfa/net.stabyourself.nottetris2.yml#L5) and [`net.stabyourself.nottetris2`](https://github.com/flathub/net.stabyourself.nottetris2/blob/5d9ca336d81947ab2e2d2989e35d5707d8c56dfa/net.stabyourself.nottetris2.yml#L1).

**Comment by razzeee** [comment]
You think `flatpak run /app/bin/nottetris2.sh` is better then `flatpak run net.stabyourself.nottetris2`?

**Comment by RokeJulianLockhart** [comment]
> https://github.com/flatpak/flatpak/issues/1258#issuecomment-2569224661

@razzeee, no. My comprehension of the request was that `nottetris2` would become the package's alias, not the full path. `$Env:PATH` already strips file extensions (at least, in Windows) and doesn't require that the full path be entered *unless* a conflict arises (and even then, it still picks one of the duplicates to be the default for when the relative path is entered).

**Comment by leandrokemp** [comment]
https://github.com/flatpak/flatpak/issues/1258#issuecomment-2569067466

Hmm those seem to be edge cases, that could be updated by the apps developers.

Another option would be to add a new element to the manifest, called something like "alias". Then make it required, so the next time the developer is updating the app, they'll have to add the new "alias" value.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/flatpak/flatpak/issues/1258#issuecomment-2570999261

@leandrokemp, and you could have Flathub CI verify that there are no conflicts amongst the aliases.
